File: src/leetcode/linkedlist/MergeTwoLists.java
package leetcode.linkedlist;

/**
 * @author chengzw
 * @description 合并两个排序的链表，https://leetcode-cn.com/problems/he-bing-liang-ge-pai-xu-de-lian-biao-lcof/
 * <p>
 * 输入两个递增排序的链表，合并这两个链表并使新链表中的节点仍然是递增排序的。
 * 输入：1->2->4, 1->3->4
 * 输出：1->1->2->3->4->4
 * @since 2021/8/26
 */
public class MergeTwoLists {

    /**
     * 思路：用一个新的链表添加元素
     * @param l1
     * @param l2
     * @return
     */
    public static ListNode solution1(ListNode l1, ListNode l2) {
        //创建一个新的链表来合并
        ListNode newNode = new ListNode();
        ListNode tail = newNode;

        //两个指针分别遍历两个链表
        while (l1 != null && l2 != null) {
            if (l1.val < l2.val) {
                //新链表添加元素
                tail.next = l1;
                tail = l1;
                //原链表后移
                l1 = l1.next;
            } else {
                tail.next = l2;
                tail = l2;
                l2 = l2.next;
            }
        }
        //剩余的链表全部接到原链表后
        if (l1 == null) {
            tail.next = l2;
        } else {
            tail.next = l1;
        }
        return newNode.next;
    }
}

File: go/leetcode/linkedlist/merge_two_lists.go
package linkedlist

/**
 * @description 剑指 Offer 25. 合并两个排序的链表
 * @author chengzw
 * @since 2022/5/18
 * @link https://leetcode.cn/problems/he-bing-liang-ge-pai-xu-de-lian-biao-lcof/
 */

/**
输入两个递增排序的链表， 合并这两个链表并使新链表中的节点仍然是递增排序的。

示例1：
输入： 1 - > 2 - > 4, 1 - > 3 - > 4
输出： 1 - > 1 - > 2 - > 3 - > 4 - > 4
*/

/**
思路：
1.使用两个指针遍历原本的两个链表，进行比较
2.再使用一个新链表来添加节点

时间复杂度：O(n+m)，n 和 m 分别是两个链表的长度
空间复杂度：O(1)
*/

func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
	// 遍历原本的两个链表
	p1 := l1
	p2 := l2
	newHead := new(ListNode)
	// 新链表的指针
	tail := newHead
	for p1 != nil && p2 != nil {
		if p1.Val < p2.Val {
			tmp := p1.Next
			p1.Next = nil
			tail.Next = p1
			tail = p1
			p1 = tmp
		} else {
			tmp := p2.Next
			p2.Next = nil
			tail.Next = p2
			tail = p2
			p2 = tmp
		}
	}
	// 把剩下一条链表的所有节点接到新链表上
	if p1 != nil {
		tail.Next = p1
	} else {
		tail.Next = p2
	}
	return newHead.Next
}
